### PR TITLE
OAuth2: add a nonce to the state parameter

### PR DESCRIPTION
**Pull request by zhaohuabing**
Commit Message: This PR adds a nonce to the state parameter of the oauth2 requests to mitigate the risk of CSRF.
Additional Description:
Risk Level: low
Testing: Unit and Integration test
Docs Changes:  API docs: add a new `oauth_nonce` to the `OAuth2Credentials.CookieNames` proto API.
Release Notes:  Yes
Platform Specific Features: No
[Optional Runtime guard:]
[Optional Fixes #Issue] Implement https://github.com/envoyproxy/envoy/issues/35232 https://github.com/envoyproxy/envoy/issues/29526
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

The diagram below shows how nonce works within the oauth2 flow, with the red annotations highlighting the changes introduced in this PR.
Untitled-2024-05-19-1553

### CODE REVIEW

**Comment by zhaohuabing** [comment]
The `verify_examples` ci failed because of a nonce has been added to the state parameter.

```
ERROR: curl (http://localhost:11901/login)
EXPECTED HEADER:
location: http://localhost:11902/authorize?client_id=0123456789&redirect_uri=http%3A%2F%2Flocalhost%3A11901%2Fauthorize&response_type=code&scope=user%3Aemail&state=http%3A%2F%2Flocalhost%3A11901%2Flogin
RECEIVED:
HTTP/1.1 302 Found

set-cookie: OauthNonce=1725418410440430;path=/;Max-Age=600;secure;HttpOnly

location: http://localhost:11902/authorize?client_id=0123456789&redirect_uri=http%3A%2F%2Flocalhost%3A11901%2Fauthorize&response_type=code&scope=user%3Aemail&state=url%3Dhttp%253A%252F%252Flocalhost%253A11901%252Flogin%26nonce%3D1725418410440430
```
https://github.com/envoyproxy/envoy/actions/runs/10694051264/job/29645555360

I can't fix this failure as it's in another repo: https://github.com/envoyproxy/examples. I'll submit a follow-up PR in the examples repo to fix it.

**Comment by phlax** [comment]
@zhaohuabing would you be able to fix the example first?

**Comment by zhaohuabing** [comment]
> @zhaohuabing would you be able to fix the example first?

Yes, I can, but some of the verifications in the example need to be temporarily disabled. This is because the content of the state parameter has changed, and the current OAuth2 filter's verification of the state differs from the new implementation. Once this PR is merged, I will re-enable those verifications in the examples.

**Comment by phlax** [comment]
hmm - was just looking - i guess my main concern is that the PR lands and breaks both repos

not sure if resolution is trivial - ie assuming the nonce is not predictable - might require some hackery on the `responds_with_header` fun

**Comment by zhaohuabing** [comment]
> hmm - was just looking - i guess my main concern is that the PR lands and breaks both repos
> 
> not sure if resolution is trivial - ie assuming the nonce is not predictable - might require some hackery on the `responds_with_header` fun

Yes, then nonce is not predictable. But we can change the verification to just check the state exists, not the value of the nonce, like what we currently do to the auth code.

**Comment by phlax** [comment]
could you raise the PR now, so its ready to go?

this will also need to be resolved for the fix to work - #35954

**Review by wbpcode — COMMENTED** [review]
Thanks for the contribution. A comment is added. I am not familiar to oauth2. So, may be we need to defer this code review to others. Or I need to take more time to read the oauth2 first.

/wait

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:58]: Enabling this by default enhances the security posture of the OAuth2 filter. This practice is also recommended by the OAuth2 spec: https://datatracker.ietf.org/doc/html/rfc6819#section-5.3.5.  So my preference is that we enable this since this release and don't provide an option to disable it unless there is a particular reason asking for it.

There are no side effects, as the nonce is transparent to both the end users and the applications.

Comment (wbpcode) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:58]: got it. Thanks.

**Comment by zhaohuabing** [comment]
> Thanks for the contribution. A comment is added. I am not familiar to oauth2. So, may be we need to defer this code review to others. Or I need to take more time to read the oauth2 first.
> 
> /wait

Hi @derekargueta @mattklein123 could you please help review this PR when you have a moment? I noticed that you're listed as owners of the OAuth2 filter in the CODEOWNERS file. Thank you so much for your time and assistance!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
